plugin/session_buffer.py

```diff
@@ -471,30 +480,32 @@ def update_document_link(self, new_link: DocumentLink) -> None:
 
     def do_document_diagnostic_async(self, view: sublime.View, version: int | None = None) -> None:
         mgr = self.session.manager()
-        if not mgr:
+        if not mgr or not self.has_capability("diagnosticProvider"):
             return
         if mgr.should_ignore_diagnostics(self._last_known_uri, self.session.config):
             return
         if version is None:
             version = view.change_count()
-        if self.has_capability("diagnosticProvider"):
-            if self._document_diagnostic_pending_response:
-                self.session.cancel_request(self._document_diagnostic_pending_response)
-            params: DocumentDiagnosticParams = {'textDocument': text_document_identifier(view)}
-            identifier = self.get_capability("diagnosticProvider.identifier")
-            if identifier:
-                params['identifier'] = identifier
-            result_id = self.session.diagnostics_result_ids.get(self._last_known_uri)
-            if result_id is not None:
-                params['previousResultId'] = result_id
-            self._document_diagnostic_pending_response = self.session.send_request_async(
-                Request.documentDiagnostic(params, view),
-                partial(self._on_document_diagnostic_async, version),
-                partial(self._on_document_diagnostic_error_async, version)
-            )
+        if self._document_diagnostic_pending_request:
+            if self._document_diagnostic_pending_request.version == version:
+                return
+            self.session.cancel_request(self._document_diagnostic_pending_request.request_id)
+        params: DocumentDiagnosticParams = {'textDocument': text_document_identifier(view)}
+        identifier = self.get_capability("diagnosticProvider.identifier")
+        if identifier:
+            params['identifier'] = identifier
+        result_id = self.session.diagnostics_result_ids.get(self._last_known_uri)
+        if result_id is not None:
+            params['previousResultId'] = result_id
+        request_id = self.session.send_request_async(
+            Request.documentDiagnostic(params, view),
+            partial(self._on_document_diagnostic_async, version),
+            partial(self._on_document_diagnostic_error_async, version)
+        )
+        self._document_diagnostic_pending_request = PendingDocumentDiagnosticRequest(version, request_id)
 
```

Comment:
Moved check for `self.has_capability("diagnosticProvider")` to the top which changed indentation quite a bit and made it harder to review.

Apart from that, added a check and bail for the case when requesting diagnostics for document version for which there already is a pending request. Without the added check it would cancel existing and trigger a new one which is a waste if the document hasn't changed since.